rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,507 @@
+    ic::MAX_LIVE = ::core::cmp::max(ic::MAX_LIVE, live_size);
+    ic::RECLAIMED += Bytes(old_limits.free as u64 - new_limits.free as u64);
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum Strategy {
+    Young,
+    Full,
+}
+
+#[cfg(feature = "ic")]
+static mut OLD_GENERATION_THRESHOLD: usize = 32 * 1024 * 1024;
+
+#[cfg(feature = "ic")]
+static mut PASSED_CRITICAL_LIMIT: bool = false;
+
+#[cfg(feature = "ic")]
+const CRITICAL_MEMORY_LIMIT: usize = (4096 - 512) * 1024 * 1024;
+
+#[cfg(feature = "ic")]
+unsafe fn decide_strategy(limits: &Limits) -> Option<Strategy> {
+    const YOUNG_GENERATION_THRESHOLD: usize = 8 * 1024 * 1024;
+
+    assert!(limits.base <= limits.last_free);
+    let old_generation_size = limits.last_free - limits.base;
+    assert!(limits.last_free <= limits.free);
+    let young_generation_size = limits.free - limits.last_free;
+
+    if limits.free >= CRITICAL_MEMORY_LIMIT && !PASSED_CRITICAL_LIMIT {
+        PASSED_CRITICAL_LIMIT = true;
+        Some(Strategy::Full)
+    } else if old_generation_size > OLD_GENERATION_THRESHOLD {
+        Some(Strategy::Full)
+    } else if young_generation_size > YOUNG_GENERATION_THRESHOLD {
+        Some(Strategy::Young)
+    } else {
+        None
+    }
+}
+
+#[cfg(feature = "ic")]
+unsafe fn update_strategy(strategy: Strategy, limits: &Limits) {
+    const GROWTH_RATE: f64 = 2.0;
+    if strategy == Strategy::Full {
+        OLD_GENERATION_THRESHOLD = ((limits.free - limits.base) as f64 * GROWTH_RATE) as usize;
+        if limits.free < CRITICAL_MEMORY_LIMIT {
+            PASSED_CRITICAL_LIMIT = false
+        }
+    }
+}
+
+pub struct Heap<'a, M: Memory> {
+    pub mem: &'a mut M,
+    pub limits: Limits,
+    pub roots: Roots,
+}
+
+pub struct Roots {
+    pub static_roots: Value,
+    pub continuation_table_ptr_loc: *mut Value,
+    // For possible future additional roots, please extend the functionality in:
+    // * `mark_root_set`
+    // * `thread_initial_phase`
+}
+
+pub struct Limits {
+    pub base: usize,
+    pub last_free: usize, // this separates the old generation from the young generation
+    pub free: usize,
+}
+
+pub struct GenerationalGC<'a, M: Memory> {
+    pub heap: Heap<'a, M>,
+    marked_space: usize,
+    strategy: Strategy,
+}
+
+impl<'a, M: Memory> GenerationalGC<'a, M> {
+    pub fn new(heap: Heap<M>, strategy: Strategy) -> GenerationalGC<M> {
+        GenerationalGC {
+            heap,
+            marked_space: 0,
+            strategy,
+        }
+    }
+
+    pub unsafe fn run(&mut self) {
+        self.alloc_mark_structures();
+        self.mark_phase();
+        self.compact_phase();
+        self.free_mark_structures();
+    }
+
+    unsafe fn alloc_mark_structures(&mut self) {
+        const BITMAP_ALIGNMENT: usize = 8 * WORD_SIZE as usize;
+        let mark_offset = match self.strategy {
+            Strategy::Young => self.heap.limits.last_free / BITMAP_ALIGNMENT * BITMAP_ALIGNMENT,
```

Comment:
Curious - do we not have the invariant that last_free is word aligned?

Reply:
Good question: It should be guaranteed because `Memory.allow_words()` is the primitive operation for heap allocations, and it uses words as the unit to allocate.